tools/bwa.py

```diff
@@ -47,7 +48,155 @@ def index(self, inFasta, algorithm=None):
+
+            TODO: With the addition of a third aligner to viral-ngs, the functionality
+            common to this method and to the comparable method in the Novoalign wrapper should
+            be broken out as an "aligner" superclass, capable of aligning bam or fastq files with an arbitrary
+            aligner, while preserving read groups. 
+        """
+        options = options or []
+
+        samtools = tools.samtools.SamtoolsTool()
+
+        # Require exactly one RG
+        rgs = samtools.getReadGroups(inBam)
+        if len(rgs) == 0:
+            raise InvalidBamHeaderError("{} lacks read groups".format(inBam))
+        elif len(rgs) == 1:
+            if not rgid:
+                rgid = list(rgs.keys())[0]
+        elif not rgid:
+            raise InvalidBamHeaderError("{} has {} read groups, but we require exactly one".format(inBam, len(rgs)))
+        if rgid not in rgs:
+            raise InvalidBamHeaderError("{} has read groups, but not {}".format(inBam, rgid))
+
+        headerFile = util.file.mkstempfname('.{}.header.txt'.format(rgid))
+        # Strip inBam to just one RG (if necessary)
+        if len(rgs) == 1:
+            one_rg_inBam = inBam
+            tools.samtools.SamtoolsTool().dumpHeader(one_rg_inBam, headerFile)
+        else:
+            # strip inBam to one read group
+            tmp_bam = util.file.mkstempfname('.onebam.bam')
+            samtools.view(['-b', '-r', rgid], inBam, tmp_bam)
+            # special exit if this file is empty
+            if samtools.count(tmp_bam) == 0:
+                return
+            # simplify BAM header otherwise Novoalign gets confused
+            one_rg_inBam = util.file.mkstempfname('.{}.in.bam'.format(rgid))
+            
+            with open(headerFile, 'wt') as outf:
+                for row in samtools.getHeader(inBam):
+                    if len(row) > 0 and row[0] == '@RG':
+                        if rgid != list(x[3:] for x in row if x.startswith('ID:'))[0]:
+                            # skip all read groups that are not rgid
+                            continue
+                    outf.write('\t'.join(row) + '\n')
+            samtools.reheader(tmp_bam, headerFile, one_rg_inBam)
+            os.unlink(tmp_bam)
+
+        # perform actual alignment
+
+        # get the read group line to give to BWA
+        readgroup_line = ""
+        with open(headerFile) as inf:
+            for line in inf:
+                if line.startswith("@RG"):
+                    readgroup_line = line
+
+        assert len(readgroup_line) > 0
+        
+        aln_bam_prefilter = util.file.mkstempfname('.prefiltered.bam')
+        # rather than reheader the alignment bam file later so it has the readgroup information
+        # from the original bam file, we'll pass the RG line to bwa to write out
+        self.mem(one_rg_inBam, refDb, aln_bam_prefilter, opts=options+['-R', readgroup_line.rstrip("\n").rstrip("\r")], min_qual=min_qual, threads=threads)
```

Comment:
Samtools' `reheader` is particularly fast when it segfaults 😕, which I've seen on BWA output with our one-RG header. I'm trying to figure out why, but others online have reported similar failures.

A `.bam` file demuxed by our process and selected at random from DNAnexus included the standard `@HD` header line and then `@RG`:
```
@HD     VN:1.4  SO:queryname
@RG     ID:ARJYU.1      DT:2016-08-05T00:00:00+0000     PU:ARJYU.1.AGGCAGAA-GCGTAAGA    LB:BWH001       SM:BWH001       CN:M04004       PL:illumina
```
According to the [spec](https://samtools.github.io/hts-specs/SAMv1.pdf), `@HD` will always be present, `@SQ` will be present for an aligned sam/bam, we can assume to have `@RG`, the `@PG` program info line may be present (do we care about it?), and a `@CO` comment may be present. Along with passing in an `@RG` line to BWA via `-R` we can pass in the other lines via `-H`, which accept an `@line`, or a set of lines from a file (we'd have to remove `@HD` for the file option).